File: src/data/courses/dsa/trees/path-relationship-problems/count-paths-with-sum/code/python/count_paths_with_sum.py
# ============================================================
# Path Sum III
# ============================================================
from collections import defaultdict
class Solution:
    def pathSum(self, root, targetSum: int) -> int:
        prefix_map = defaultdict(int)
        prefix_map[0] = 1

        def dfs(node, current_sum):
            if not node: return 0
            current_sum += node.val
            count = prefix_map[current_sum - targetSum]
            prefix_map[current_sum] += 1
            count += dfs(node.left, current_sum) + dfs(node.right, current_sum)
            prefix_map[current_sum] -= 1  # Backtrack
            return count

        return dfs(root, 0)

File: src/data/courses/dsa/trees/path-relationship-problems/count-paths-with-sum/code/java/CountPathsWithSum.java
// ============================================================
// Path Sum III
// ============================================================
import java.util.*;
class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        Map<Long, Integer> map = new HashMap<>();
        map.put(0L, 1);
        return dfs(root, 0, targetSum, map);
    }
    int dfs(TreeNode node, long curSum, int target, Map<Long, Integer> map) {
        if (node == null) return 0;
        curSum += node.val;
        int count = map.getOrDefault(curSum - target, 0);
        map.merge(curSum, 1, Integer::sum);
        count += dfs(node.left, curSum, target, map);
        count += dfs(node.right, curSum, target, map);
        map.merge(curSum, -1, Integer::sum); // Backtrack
        return count;
    }
}
